### PR TITLE
new split reduce heuristic try 2

### PR DESCRIPTION
**Pull request by chaosagent**
#3780 

don't know what happened to the hcopt slowdown, seems fine now? (tested gpt-2 and SD)

2**22 output buffer is a little aggressively high and may be overfit on nvidia resnet. Maybe we can back it off when we get reduce-padto+tc and group+tc

- can there be scheduler improvements to make these two kernels sequential to take advantage of l2/l3 cache?
- what is the performance diff on amd vs nvidia?

### CODE REVIEW

**Comment by chaosagent** [comment]
benchmarks!


```
(venv) chaos@tiny18:~/tinygrad$ SPLIT_OUT_MAX=23 JITCNT=20 BEAM=4 CUDA_VISIBLE_DEVICES=1 DEFAULT_FLOAT=HALF BS=256 PARALLEL=10 python test/external/external_benchmark_resnet.py
layer1 slice1 x(256, 64, 56, 56)          :     34.47 ms,   11.23 tflops,    4.49 GB used,    112 kernels
layer1 slice2 x(256, 256, 56, 56)         :     29.90 ms,   11.72 tflops,    5.70 GB used,     87 kernels
layer2 slice1 x(256, 256, 56, 56)         :     39.86 ms,   14.76 tflops,    4.94 GB used,    112 kernels
layer2 slice2 x(256, 512, 28, 28)         :     19.77 ms,   17.35 tflops,    2.89 GB used,     87 kernels
layer3 slice1 x(256, 512, 28, 28)         :     27.17 ms,   21.36 tflops,    2.54 GB used,    112 kernels
layer3 slice2 x(256, 1024, 14, 14)        :     15.09 ms,   22.49 tflops,    1.49 GB used,     87 kernels
layer4 slice1 x(256, 1024, 14, 14)        :     24.70 ms,   23.34 tflops,    1.32 GB used,     89 kernels
layer4 slice2 x(256, 2048, 7, 7)          :     11.92 ms,   28.31 tflops,    0.76 GB used,     63 kernels
estimated step tm: 344.57 ms, 18.089 tflops, 442.98 GB/s, 1421 kernels

----------------------------------------------------------------------
Ran 8 tests in 83.530s

OK

(venv) chaos@tiny18:~/tinygrad$ SPLIT_OUT_MAX=22 JITCNT=20 BEAM=4 CUDA_VISIBLE_DEVICES=1 DEFAULT_FLOAT=HALF BS=256 PARALLEL=10 python test/external/external_benchmark_resnet.py
layer1 slice1 x(256, 64, 56, 56)          :     35.06 ms,   11.05 tflops,    4.48 GB used,    112 kernels
layer1 slice2 x(256, 256, 56, 56)         :     29.85 ms,   11.73 tflops,    5.69 GB used,     87 kernels
layer2 slice1 x(256, 256, 56, 56)         :     41.41 ms,   14.21 tflops,    4.90 GB used,    112 kernels
layer2 slice2 x(256, 512, 28, 28)         :     19.79 ms,   17.33 tflops,    2.87 GB used,     87 kernels
layer3 slice1 x(256, 512, 28, 28)         :     27.89 ms,   20.81 tflops,    2.49 GB used,    111 kernels
layer3 slice2 x(256, 1024, 14, 14)        :     14.07 ms,   24.11 tflops,    1.45 GB used,     86 kernels
layer4 slice1 x(256, 1024, 14, 14)        :     22.35 ms,   25.80 tflops,    1.30 GB used,     89 kernels
layer4 slice2 x(256, 2048, 7, 7)          :     11.91 ms,   28.33 tflops,    0.76 GB used,     63 kernels
estimated step tm: 339.94 ms, 18.335 tflops, 445.69 GB/s, 1415 kernels

----------------------------------------------------------------------
Ran 8 tests in 82.990s

OK
(venv) chaos@tiny18:~/tinygrad$ SPLIT_OUT_MAX=21 JITCNT=20 BEAM=4 CUDA_VISIBLE_DEVICES=1 DEFAULT_FLOAT=HALF BS=256 PARALLEL=10 python test/external/external_benchmark_resnet.py
layer1 slice1 x(256, 64, 56, 56)          :     36.46 ms,   10.62 tflops,    4.45 GB used,    112 kernels
layer1 slice2 x(256, 256, 56, 56)         :     32.51 ms,   10.77 tflops,    5.67 GB used,     87 kernels
layer2 slice1 x(256, 256, 56, 56)         :     37.85 ms,   15.54 tflops,    4.87 GB used,    112 kernels
layer2 slice2 x(256, 512, 28, 28)         :     19.87 ms,   17.26 tflops,    2.85 GB used,     87 kernels
layer3 slice1 x(256, 512, 28, 28)         :     29.77 ms,   19.49 tflops,    2.46 GB used,    110 kernels
layer3 slice2 x(256, 1024, 14, 14)        :     14.36 ms,   23.62 tflops,    1.45 GB used,     86 kernels
layer4 slice1 x(256, 1024, 14, 14)        :     24.57 ms,   23.46 tflops,    1.29 GB used,     88 kernels
layer4 slice2 x(256, 2048, 7, 7)          :     11.91 ms,   28.35 tflops,    0.76 GB used,     63 kernels
estimated step tm: 348.90 ms, 17.864 tflops, 432.57 GB/s, 1413 kernels

----------------------------------------------------------------------
Ran 8 tests in 82.158s

OK
(venv) chaos@tiny18:~/tinygrad$ SPLIT_OUT_MAX=20 JITCNT=20 BEAM=4 CUDA_VISIBLE_DEVICES=1 DEFAULT_FLOAT=HALF BS=256 PARALLEL=10 python test/external/external_benchmark_resnet.py
layer1 slice1 x(256, 64, 56, 56)          :     35.05 ms,   11.05 tflops,    4.44 GB used,    112 kernels
layer1 slice2 x(256, 256, 56, 56)         :     30.59 ms,   11.45 tflops,    5.67 GB used,     87 kernels
layer2 slice1 x(256, 256, 56, 56)         :     37.16 ms,   15.83 tflops,    4.86 GB used,    111 kernels
layer2 slice2 x(256, 512, 28, 28)         :     21.00 ms,   16.33 tflops,    2.84 GB used,     86 kernels
layer3 slice1 x(256, 512, 28, 28)         :     29.75 ms,   19.50 tflops,    2.46 GB used,    109 kernels
layer3 slice2 x(256, 1024, 14, 14)        :     16.10 ms,   21.07 tflops,    1.44 GB used,     84 kernels
layer4 slice1 x(256, 1024, 14, 14)        :     24.60 ms,   23.43 tflops,    1.29 GB used,     88 kernels
layer4 slice2 x(256, 2048, 7, 7)          :     11.88 ms,   28.42 tflops,    0.76 GB used,     63 kernels
estimated step tm: 354.97 ms, 17.558 tflops, 424.12 GB/s, 1398 kernels

----------------------------------------------------------------------
Ran 8 tests in 83.866s

OK
(venv) chaos@tiny18:~/tinygrad$ SPLIT_OUT_MAX=19 JITCNT=20 BEAM=4 CUDA_VISIBLE_DEVICES=1 DEFAULT_FLOAT=HALF BS=256 PARALLEL=10 python test/external/external_benchmark_resnet.py
layer1 slice1 x(256, 64, 56, 56)          :     38.97 ms,    9.94 tflops,    4.43 GB used,    112 kernels
layer1 slice2 x(256, 256, 56, 56)         :     32.47 ms,   10.79 tflops,    5.66 GB used,     87 kernels
layer2 slice1 x(256, 256, 56, 56)         :     38.94 ms,   15.11 tflops,    4.85 GB used,    110 kernels
layer2 slice2 x(256, 512, 28, 28)         :     20.60 ms,   16.65 tflops,    2.84 GB used,     86 kernels
layer3 slice1 x(256, 512, 28, 28)         :     30.55 ms,   19.00 tflops,    2.45 GB used,    108 kernels
layer3 slice2 x(256, 1024, 14, 14)        :     16.07 ms,   21.10 tflops,    1.44 GB used,     84 kernels
layer4 slice1 x(256, 1024, 14, 14)        :     24.72 ms,   23.33 tflops,    1.29 GB used,     88 kernels
layer4 slice2 x(256, 2048, 7, 7)          :     11.89 ms,   28.39 tflops,    0.76 GB used,     63 kernels
estimated step tm: 364.03 ms, 17.121 tflops, 413.33 GB/s, 1396 kernels

----------------------------------------------------------------------
Ran 8 tests in 81.148s

OK
```

**Comment by chaosagent** [comment]
This doesn't seem to hurt anything. Run update_benchmark?

**Comment by chenyuxyz** [comment]
before i run it, can you change the resnet training benchmark to use 256 for 1 gpu, 1536 for 6 gpus and HALF and verify that memory is good without split?

**Comment by chenyuxyz** [comment]
ok i am convinced this is a much better heuristic. improved cifar speed and memory by 40% and gpt2 mac by 10+%. speed vs torch on sum on cuda also doubled

**Comment by chaosagent** [comment]
seems better than SPLIT_REDUCEOP=0 on all layers, maybe some pathological behavior? 

```
(venv) chaos@tiny18:~/tinygrad$ SPLIT_REDUCEOP=0 JITCNT=20 BEAM=4 CUDA_VISIBLE_DEVICES=1 DEFAULT_FLOAT=HALF BS=256 PARALLEL=10 python test/external/external_benchmark_resnet.py
layer1 slice1 x(256, 64, 56, 56)          :     43.65 ms,    8.83 tflops,    4.43 GB used,     87 kernels
layer1 slice2 x(256, 256, 56, 56)         :     38.46 ms,    9.15 tflops,    5.66 GB used,     63 kernels
layer2 slice1 x(256, 256, 56, 56)         :     42.25 ms,   13.98 tflops,    4.86 GB used,     88 kernels
layer2 slice2 x(256, 512, 28, 28)         :     22.26 ms,   15.44 tflops,    2.84 GB used,     67 kernels
layer3 slice1 x(256, 512, 28, 28)         :     30.29 ms,   19.19 tflops,    2.46 GB used,     97 kernels
layer3 slice2 x(256, 1024, 14, 14)        :     16.65 ms,   20.39 tflops,    1.43 GB used,     67 kernels
layer4 slice1 x(256, 1024, 14, 14)        :     25.74 ms,   22.41 tflops,    1.30 GB used,     97 kernels
layer4 slice2 x(256, 2048, 7, 7)          :     12.05 ms,   28.02 tflops,    0.76 GB used,     67 kernels
estimated step tm: 392.99 ms, 15.882 tflops, 382.11 GB/s, 1165 kernels

----------------------------------------------------------------------
Ran 8 tests in 2756.683s

OK
```
